File: storage/src/tuple.rs
//! Heap tuple implementation.
//!
//! This module implements PostgreSQL's heap tuple format as defined in
//! `src/include/access/htup_details.h`. A heap tuple consists of:
//!
//! ```text
//! +------------------+------------+-------------+
//! | HeapTupleHeader  | null bitmap| user data   |
//! | (23 bytes + pad) | (optional) | (attributes)|
//! +------------------+------------+-------------+
//! ```
//!
//! The header contains MVCC information (xmin, xmax, cid) and metadata
//! about the tuple's structure.

use bytemuck::{Pod, Zeroable};
use static_assertions::const_assert_eq;

use crate::types::{CommandId, OffsetNumber, TransactionId};

/// Size of the heap tuple header in bytes (23 bytes, but aligned to 24).
///
/// PostgreSQL's `HeapTupleHeaderData` is 23 bytes, but due to alignment
/// requirements, the actual header size is typically 24 bytes.
pub const HEAP_TUPLE_HEADER_SIZE: usize = 24;

/// Minimum heap tuple header size (without alignment padding).
pub const MIN_HEAP_TUPLE_HEADER_SIZE: usize = 23;

/// Heap tuple header data structure.
///
/// Matches PostgreSQL's `HeapTupleHeaderData` from `htup_details.h`:
///
/// ```c
/// typedef struct HeapTupleHeaderData {
///     union {
///         HeapTupleFields t_heap;
///         DatumTupleFields t_datum;
///     } t_choice;
///     ItemPointerData t_ctid;      /* current TID of this or newer tuple */
///     uint16 t_infomask2;          /* number of attributes + flags */
///     uint16 t_infomask;           /* various flag bits */
///     uint8 t_hoff;                /* sizeof header incl. bitmap, padding */
///     /* ^ - 23 bytes - ^ */
///     bits8 t_bits[FLEXIBLE_ARRAY_MEMBER]; /* bitmap of NULLs */
/// } HeapTupleHeaderData;
/// ```
///
/// We use the heap tuple variant (`t_heap`) which contains:
/// - `t_xmin`: inserting transaction ID
/// - `t_xmax`: deleting transaction ID (or 0)
/// - `t_cid`: command ID (or combo command ID)
#[derive(Debug, Clone, Copy, Pod, Zeroable)]
#[repr(C)]
pub struct HeapTupleHeader {
    /// Inserting transaction ID.
    pub t_xmin: TransactionId,
    /// Deleting transaction ID, or 0 if not deleted.
    pub t_xmax: TransactionId,
    /// Command ID within the transaction, or combo command ID.
    ///
    /// In PostgreSQL, this field is overloaded:
    /// - For regular tuples: command ID of inserting command
    /// - For updated tuples: can be a "combo" CID
    pub t_cid: CommandId,
    /// Current TID of this or newer tuple version (for HOT chains).
    ///
    /// CTID block number.
    pub t_ctid_block: u32,
    /// CTID offset number.
    pub t_ctid_offset: OffsetNumber,
    /// Number of attributes + various flag bits.
    ///
    /// Low 11 bits: number of attributes
    /// High 5 bits: flags (see `Infomask2Flags`)
    pub t_infomask2: u16,
    /// Various flag bits (see `InfomaskFlags`).
    pub t_infomask: u16,
    /// Size of header including bitmap and padding.
    pub t_hoff: u8,
    /// Padding to align to 24 bytes.
    padding: u8,
}

const_assert_eq!(std::mem::size_of::<HeapTupleHeader>(), HEAP_TUPLE_HEADER_SIZE);

/// Maximum number of attributes in a tuple.
pub const MAX_HEAP_ATTRIBUTE_NUMBER: u16 = 1600;

/// Mask for extracting attribute count from `t_infomask2`.
const HEAP_NATTS_MASK: u16 = 0x07FF;

bitflags::bitflags! {
    /// Infomask flags (t_infomask field).
    ///
    /// These flags indicate the state of the tuple for MVCC and other purposes.
    #[derive(Debug, Clone, Copy, PartialEq, Eq)]
    pub struct InfomaskFlags: u16 {
        /// Tuple has null attribute(s).
        const HEAP_HASNULL = 0x0001;
        /// Tuple has variable-width attribute(s).
        const HEAP_HASVARWIDTH = 0x0002;
        /// Tuple has external stored attribute(s).
        const HEAP_HASEXTERNAL = 0x0004;
        /// Tuple has an object ID field.
        const HEAP_HASOID_OLD = 0x0008;
        /// xmax is a key-share locker.
        const HEAP_XMAX_KEYSHR_LOCK = 0x0010;
        /// t_cid is a combo CID.
        const HEAP_COMBOCID = 0x0020;
        /// xmax is exclusive locker.
        const HEAP_XMAX_EXCL_LOCK = 0x0040;
        /// xmax, if valid, is only a locker.
        const HEAP_XMAX_LOCK_ONLY = 0x0080;
        /// xmin committed.
        const HEAP_XMIN_COMMITTED = 0x0100;
        /// xmin invalid/aborted.
        const HEAP_XMIN_INVALID = 0x0200;
        /// xmax committed.
        const HEAP_XMAX_COMMITTED = 0x0400;
        /// xmax invalid/aborted.
        const HEAP_XMAX_INVALID = 0x0800;
        /// xmax is a MultiXactId.
        const HEAP_XMAX_IS_MULTI = 0x1000;
        /// tuple was updated.
        const HEAP_UPDATED = 0x2000;
        /// tuple moved to another place by pre-9.0 VACUUM FULL.
        const HEAP_MOVED_OFF = 0x4000;
        /// tuple moved in from another place by pre-9.0 VACUUM FULL.
        const HEAP_MOVED_IN = 0x8000;
    }
}

bitflags::bitflags! {
    /// Infomask2 flags (high bits of t_infomask2 field).
    #[derive(Debug, Clone, Copy, PartialEq, Eq)]
    pub struct Infomask2Flags: u16 {
        /// Tuple has HOT successor.
        const HEAP_HOT_UPDATED = 0x4000;
        /// Tuple is HOT-updated version.
        const HEAP_ONLY_TUPLE = 0x8000;
    }
}

impl HeapTupleHeader {
    /// Create a new heap tuple header.
    #[must_use]
    pub fn new(xmin: TransactionId, natts: u16) -> Self {
        Self {
            t_xmin: xmin,
            t_xmax: 0,
            t_cid: 0,
            t_ctid_block: 0,
            t_ctid_offset: 0,
            t_infomask2: natts & HEAP_NATTS_MASK,
            t_infomask: InfomaskFlags::HEAP_XMAX_INVALID.bits(),
            t_hoff: HEAP_TUPLE_HEADER_SIZE as u8,
            padding: 0,
        }
    }

    /// Get the number of attributes.
    #[must_use]
    pub const fn natts(&self) -> u16 {
        self.t_infomask2 & HEAP_NATTS_MASK
    }

    /// Set the number of attributes.
    pub fn set_natts(&mut self, natts: u16) {
        self.t_infomask2 = (self.t_infomask2 & !HEAP_NATTS_MASK) | (natts & HEAP_NATTS_MASK);
    }

    /// Get the infomask flags.
    #[must_use]
    pub fn infomask(&self) -> InfomaskFlags {
        InfomaskFlags::from_bits_truncate(self.t_infomask)
    }

    /// Set the infomask flags.
    pub fn set_infomask(&mut self, flags: InfomaskFlags) {
        self.t_infomask = flags.bits();
    }

    /// Get the infomask2 flags (excluding attribute count).
    #[must_use]
    pub fn infomask2_flags(&self) -> Infomask2Flags {
        Infomask2Flags::from_bits_truncate(self.t_infomask2)
    }

    /// Check if the tuple has null attributes.
    #[must_use]
    pub fn has_nulls(&self) -> bool {
        self.infomask().contains(InfomaskFlags::HEAP_HASNULL)
    }

    /// Check if xmin is committed.
    #[must_use]
    pub fn xmin_committed(&self) -> bool {
        self.infomask().contains(InfomaskFlags::HEAP_XMIN_COMMITTED)
    }

    /// Check if xmin is invalid/aborted.
    #[must_use]
    pub fn xmin_invalid(&self) -> bool {
        self.infomask().contains(InfomaskFlags::HEAP_XMIN_INVALID)
    }

    /// Check if xmax is committed.
    #[must_use]
    pub fn xmax_committed(&self) -> bool {
        self.infomask().contains(InfomaskFlags::HEAP_XMAX_COMMITTED)
    }

    /// Check if xmax is invalid/aborted.
    #[must_use]
    pub fn xmax_invalid(&self) -> bool {
        self.infomask().contains(InfomaskFlags::HEAP_XMAX_INVALID)
    }

    /// Check if this is a HOT-updated tuple.
    #[must_use]
    pub fn is_hot_updated(&self) -> bool {
        self.infomask2_flags().contains(Infomask2Flags::HEAP_HOT_UPDATED)
    }

    /// Check if this is a heap-only tuple (HOT).
    #[must_use]
    pub fn is_heap_only(&self) -> bool {
        self.infomask2_flags().contains(Infomask2Flags::HEAP_ONLY_TUPLE)
    }

    /// Mark xmin as committed.
    pub fn set_xmin_committed(&mut self) {
        let mut flags = self.infomask();
        flags.insert(InfomaskFlags::HEAP_XMIN_COMMITTED);
        flags.remove(InfomaskFlags::HEAP_XMIN_INVALID);
        self.set_infomask(flags);
    }

    /// Mark xmin as invalid/aborted.
    pub fn set_xmin_invalid(&mut self) {
        let mut flags = self.infomask();
        flags.insert(InfomaskFlags::HEAP_XMIN_INVALID);
        flags.remove(InfomaskFlags::HEAP_XMIN_COMMITTED);
        self.set_infomask(flags);
    }

    /// Mark xmax as committed.
    pub fn set_xmax_committed(&mut self) {
        let mut flags = self.infomask();
        flags.insert(InfomaskFlags::HEAP_XMAX_COMMITTED);
        flags.remove(InfomaskFlags::HEAP_XMAX_INVALID);
        self.set_infomask(flags);
    }

    /// Mark xmax as invalid/aborted.
    pub fn set_xmax_invalid(&mut self) {
        let mut flags = self.infomask();
        flags.insert(InfomaskFlags::HEAP_XMAX_INVALID);
        flags.remove(InfomaskFlags::HEAP_XMAX_COMMITTED);
        self.set_infomask(flags);
    }

    /// Set the CTID (current tuple ID).
    pub fn set_ctid(&mut self, block: u32, offset: OffsetNumber) {
        self.t_ctid_block = block;
        self.t_ctid_offset = offset;
    }

    /// Get the CTID as (block, offset).
    #[must_use]
    pub const fn ctid(&self) -> (u32, OffsetNumber) {
        (self.t_ctid_block, self.t_ctid_offset)
    }
}

impl Default for HeapTupleHeader {
    fn default() -> Self {
        Self::new(0, 0)
    }
}

/// A complete heap tuple with header and data.
#[derive(Debug, Clone)]
pub struct HeapTuple {
    /// The tuple header.
    pub header: HeapTupleHeader,
    /// The tuple data (null bitmap + attribute values).
    pub data: Vec<u8>,
}

impl HeapTuple {
    /// Create a new heap tuple.
    #[must_use]
    pub fn new(xmin: TransactionId, natts: u16, data: Vec<u8>) -> Self {
        Self {
            header: HeapTupleHeader::new(xmin, natts),
            data,
        }
    }

    /// Get the total size of the tuple (header + data).
    #[must_use]
    pub fn len(&self) -> usize {
        HEAP_TUPLE_HEADER_SIZE + self.data.len()
    }

    /// Check if the tuple has no data.
    #[must_use]
    pub fn is_empty(&self) -> bool {
        self.data.is_empty()
    }

    /// Serialize the tuple to bytes.
    #[must_use]
    pub fn to_bytes(&self) -> Vec<u8> {
        let mut bytes = Vec::with_capacity(self.len());
        bytes.extend_from_slice(bytemuck::bytes_of(&self.header));
        bytes.extend_from_slice(&self.data);
        bytes
    }

    /// Deserialize a tuple from bytes.
    ///
    /// # Panics
    ///
    /// Panics if the bytes are too short to contain a header.
    #[must_use]
    pub fn from_bytes(bytes: &[u8]) -> Self {
        assert!(
            bytes.len() >= HEAP_TUPLE_HEADER_SIZE,
            "bytes too short for tuple header"
        );
        let header: HeapTupleHeader = *bytemuck::from_bytes(&bytes[..HEAP_TUPLE_HEADER_SIZE]);
        let data = bytes[HEAP_TUPLE_HEADER_SIZE..].to_vec();
        Self { header, data }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn header_size_is_correct() {
        assert_eq!(std::mem::size_of::<HeapTupleHeader>(), HEAP_TUPLE_HEADER_SIZE);
    }

    #[test]
    fn new_header_defaults() {
        let header = HeapTupleHeader::new(100, 5);
        assert_eq!(header.t_xmin, 100);
        assert_eq!(header.t_xmax, 0);
        assert_eq!(header.natts(), 5);
        assert!(header.xmax_invalid());
        assert!(!header.xmin_committed());
    }

    #[test]
    fn natts_encoding() {
        let mut header = HeapTupleHeader::new(0, 0);

        header.set_natts(100);
        assert_eq!(header.natts(), 100);

        header.set_natts(MAX_HEAP_ATTRIBUTE_NUMBER);
        assert_eq!(header.natts(), MAX_HEAP_ATTRIBUTE_NUMBER & HEAP_NATTS_MASK);
    }

    #[test]
    fn infomask_operations() {
        let mut header = HeapTupleHeader::new(100, 5);

        assert!(!header.xmin_committed());
        header.set_xmin_committed();
        assert!(header.xmin_committed());
        assert!(!header.xmin_invalid());

        header.set_xmin_invalid();
        assert!(header.xmin_invalid());
        assert!(!header.xmin_committed());
    }

    #[test]
    fn ctid_operations() {
        let mut header = HeapTupleHeader::new(100, 5);

        header.set_ctid(42, 7);
        assert_eq!(header.ctid(), (42, 7));
    }

    #[test]
    fn tuple_serialization() {
        let tuple = HeapTuple::new(100, 3, vec![1, 2, 3, 4, 5]);

        let bytes = tuple.to_bytes();
        assert_eq!(bytes.len(), HEAP_TUPLE_HEADER_SIZE + 5);

        let restored = HeapTuple::from_bytes(&bytes);
        assert_eq!(restored.header.t_xmin, 100);
        assert_eq!(restored.header.natts(), 3);
        assert_eq!(restored.data, vec![1, 2, 3, 4, 5]);
    }

    #[test]
    fn tuple_length() {
        let tuple = HeapTuple::new(0, 0, vec![0; 100]);
        assert_eq!(tuple.len(), HEAP_TUPLE_HEADER_SIZE + 100);
    }
}

File: internal/storage/tuple.go
package storage

// Heap tuple layout (from src/include/access/htup_details.h):
//
//	+------------------+------------+-------------+
//	| HeapTupleHeader  | null bitmap| user data   |
//	| (23 bytes + pad) | (optional) | (attributes)|
//	+------------------+------------+-------------+
//
// The header contains MVCC fields (xmin, xmax, cid), a self-TID, infomask
// flags, and a header-offset field (t_hoff) pointing past any null bitmap.
//
// We model the "heap" variant of t_choice (i.e. t_heap, not t_datum).

import "encoding/binary"

// HeapTupleHeaderSize is the fixed size of HeapTupleHeader in bytes.
// PostgreSQL's HeapTupleHeaderData is 23 bytes but padded to 24.
const HeapTupleHeaderSize = 24

// MaxHeapAttributeNumber is the maximum number of attributes in a tuple.
const MaxHeapAttributeNumber uint16 = 1600

// heapNattsMask extracts the attribute count from t_infomask2 (low 11 bits).
const heapNattsMask uint16 = 0x07FF

// InfomaskFlags are bits in HeapTupleHeader.TInfomask.
// Matches PostgreSQL's HEAP_* flags in htup_details.h.
type InfomaskFlags uint16

const (
	HeapHasNull        InfomaskFlags = 0x0001 // tuple has at least one null attribute
	HeapHasVarWidth    InfomaskFlags = 0x0002 // has variable-length attribute(s)
	HeapHasExternal    InfomaskFlags = 0x0004 // has external TOAST attribute(s)
	HeapHasOidOld      InfomaskFlags = 0x0008 // has OID field (pre-12 compat)
	HeapXmaxKeyshrLock InfomaskFlags = 0x0010 // xmax is a key-share locker
	HeapComboCid       InfomaskFlags = 0x0020 // t_cid is a combo CID
	HeapXmaxExclLock   InfomaskFlags = 0x0040 // xmax is exclusive locker
	HeapXmaxLockOnly   InfomaskFlags = 0x0080 // xmax is a locker, not deleter
	HeapXminCommitted  InfomaskFlags = 0x0100 // xmin committed (hint bit)
	HeapXminInvalid    InfomaskFlags = 0x0200 // xmin invalid / aborted (hint bit)
	HeapXmaxCommitted  InfomaskFlags = 0x0400 // xmax committed (hint bit)
	HeapXmaxInvalid    InfomaskFlags = 0x0800 // xmax invalid / aborted (hint bit)
	HeapXmaxIsMulti    InfomaskFlags = 0x1000 // xmax is a MultiXactId
	HeapUpdated        InfomaskFlags = 0x2000 // tuple was updated
	HeapMovedOff       InfomaskFlags = 0x4000 // moved by pre-9.0 VACUUM FULL
	HeapMovedIn        InfomaskFlags = 0x8000 // moved in by pre-9.0 VACUUM FULL
)

// Infomask2Flags are the high bits of HeapTupleHeader.TInfomask2.
// The low 11 bits of t_infomask2 encode the attribute count.
type Infomask2Flags uint16

const (
	HeapHotUpdated Infomask2Flags = 0x4000 // tuple has HOT successor
	HeapOnlyTuple  Infomask2Flags = 0x8000 // tuple is a heap-only (HOT) version
)

// HeapTupleHeader is the 24-byte header prepended to every heap tuple.
//
// Matches PostgreSQL's HeapTupleHeaderData (htup_details.h):
//
//	typedef struct HeapTupleHeaderData {
//	    union { HeapTupleFields t_heap; DatumTupleFields t_datum; } t_choice;
//	    ItemPointerData t_ctid;     // 6 bytes
//	    uint16          t_infomask2;
//	    uint16          t_infomask;
//	    uint8           t_hoff;
//	    /* ^ 23 bytes ^ */
//	    bits8 t_bits[FLEXIBLE_ARRAY_MEMBER]; // null bitmap
//	} HeapTupleHeaderData;
//
// We use the t_heap variant:
//   - TXmin (4): inserting XID
//   - TXmax (4): deleting XID (0 if alive)
//   - TCid  (4): command ID within transaction
//   - TCtidBlock (4) + TCtidOffset (2): self/forward TID
//   - TInfomask2 (2), TInfomask (2), THoff (1), _pad (1) → total 24 bytes
type HeapTupleHeader struct {
	TXmin       TransactionId
	TXmax       TransactionId
	TCid        CommandId
	TCtidBlock  BlockNumber
	TCtidOffset OffsetNumber
	TInfomask2  uint16
	TInfomask   uint16
	THoff       uint8
	_pad        uint8 //nolint:unused // alignment padding to reach 24 bytes
}

// NewHeapTupleHeader creates a header for a freshly-inserted tuple.
// xmin is the inserting transaction; natts is the number of attributes.
// xmax is initialised to 0 with HEAP_XMAX_INVALID set.
func NewHeapTupleHeader(xmin TransactionId, natts uint16) HeapTupleHeader {
	return HeapTupleHeader{
		TXmin:      xmin,
		TXmax:      0,
		TCid:       0,
		TInfomask2: natts & heapNattsMask,
		TInfomask:  uint16(HeapXmaxInvalid),
		THoff:      HeapTupleHeaderSize,
	}
}

// Natts returns the number of attributes encoded in TInfomask2.
func (h *HeapTupleHeader) Natts() uint16 { return h.TInfomask2 & heapNattsMask }

// SetNatts sets the attribute count in TInfomask2, preserving flag bits.
func (h *HeapTupleHeader) SetNatts(natts uint16) {
	h.TInfomask2 = (h.TInfomask2 & ^heapNattsMask) | (natts & heapNattsMask)
}

// Infomask returns the InfomaskFlags.
func (h *HeapTupleHeader) Infomask() InfomaskFlags {
	return InfomaskFlags(h.TInfomask)
}

// SetInfomask replaces the infomask.
func (h *HeapTupleHeader) SetInfomask(f InfomaskFlags) { h.TInfomask = uint16(f) }

// Infomask2Flags returns the high-bit flags from TInfomask2.
func (h *HeapTupleHeader) Infomask2Flags() Infomask2Flags {
	return Infomask2Flags(h.TInfomask2 & ^heapNattsMask)
}

// --- Hint bit helpers ---

func (h *HeapTupleHeader) XminCommitted() bool {
	return h.Infomask()&HeapXminCommitted != 0
}
func (h *HeapTupleHeader) XminInvalid() bool {
	return h.Infomask()&HeapXminInvalid != 0
}
func (h *HeapTupleHeader) XmaxCommitted() bool {
	return h.Infomask()&HeapXmaxCommitted != 0
}
func (h *HeapTupleHeader) XmaxInvalid() bool {
	return h.Infomask()&HeapXmaxInvalid != 0
}
func (h *HeapTupleHeader) IsHotUpdated() bool {
	return h.Infomask2Flags()&HeapHotUpdated != 0
}
func (h *HeapTupleHeader) IsHeapOnly() bool {
	return h.Infomask2Flags()&HeapOnlyTuple != 0
}

func (h *HeapTupleHeader) SetXminCommitted() {
	h.TInfomask = uint16((h.Infomask() | HeapXminCommitted) &^ HeapXminInvalid)
}
func (h *HeapTupleHeader) SetXminInvalid() {
	h.TInfomask = uint16((h.Infomask() | HeapXminInvalid) &^ HeapXminCommitted)
}
func (h *HeapTupleHeader) SetXmaxCommitted() {
	h.TInfomask = uint16((h.Infomask() | HeapXmaxCommitted) &^ HeapXmaxInvalid)
}
func (h *HeapTupleHeader) SetXmaxInvalid() {
	h.TInfomask = uint16((h.Infomask() | HeapXmaxInvalid) &^ HeapXmaxCommitted)
}

// Ctid returns the self-referential TID as (BlockNumber, OffsetNumber).
func (h *HeapTupleHeader) Ctid() (BlockNumber, OffsetNumber) {
	return h.TCtidBlock, h.TCtidOffset
}

// SetCtid sets the self TID.
func (h *HeapTupleHeader) SetCtid(block BlockNumber, offset OffsetNumber) {
	h.TCtidBlock = block
	h.TCtidOffset = offset
}

// encodeHeapTupleHeader serializes h into dst[0:HeapTupleHeaderSize].
func encodeHeapTupleHeader(dst []byte, h *HeapTupleHeader) {
	binary.LittleEndian.PutUint32(dst[0:], h.TXmin)
	binary.LittleEndian.PutUint32(dst[4:], h.TXmax)
	binary.LittleEndian.PutUint32(dst[8:], h.TCid)
	binary.LittleEndian.PutUint32(dst[12:], h.TCtidBlock)
	binary.LittleEndian.PutUint16(dst[16:], h.TCtidOffset)
	binary.LittleEndian.PutUint16(dst[18:], h.TInfomask2)
	binary.LittleEndian.PutUint16(dst[20:], h.TInfomask)
	dst[22] = h.THoff
	dst[23] = 0 // padding
}

// decodeHeapTupleHeader deserializes from src[0:HeapTupleHeaderSize].
func decodeHeapTupleHeader(src []byte) HeapTupleHeader {
	return HeapTupleHeader{
		TXmin:       binary.LittleEndian.Uint32(src[0:]),
		TXmax:       binary.LittleEndian.Uint32(src[4:]),
		TCid:        binary.LittleEndian.Uint32(src[8:]),
		TCtidBlock:  binary.LittleEndian.Uint32(src[12:]),
		TCtidOffset: binary.LittleEndian.Uint16(src[16:]),
		TInfomask2:  binary.LittleEndian.Uint16(src[18:]),
		TInfomask:   binary.LittleEndian.Uint16(src[20:]),
		THoff:       src[22],
	}
}

// HeapTuple pairs a HeapTupleHeader with its variable-length data payload.
type HeapTuple struct {
	Header HeapTupleHeader
	Data   []byte // null bitmap + attribute values
}

// NewHeapTuple creates a tuple with the given xmin, attribute count, and data.
func NewHeapTuple(xmin TransactionId, natts uint16, data []byte) *HeapTuple {
	return &HeapTuple{
		Header: NewHeapTupleHeader(xmin, natts),
		Data:   data,
	}
}

// Len returns the total byte size: header + data.
func (t *HeapTuple) Len() int { return HeapTupleHeaderSize + len(t.Data) }

// ToBytes serializes the tuple to a byte slice.
func (t *HeapTuple) ToBytes() []byte {
	buf := make([]byte, t.Len())
	encodeHeapTupleHeader(buf[:HeapTupleHeaderSize], &t.Header)
	copy(buf[HeapTupleHeaderSize:], t.Data)
	return buf
}

// HeapTupleFromBytes deserializes a tuple from bytes.
// Returns an error if len(src) < HeapTupleHeaderSize.
func HeapTupleFromBytes(src []byte) (*HeapTuple, error) {
	if len(src) < HeapTupleHeaderSize {
		return nil, &StorageError{
			Code:    ErrInvalidPageLayout,
			Message: "bytes too short for HeapTupleHeader",
		}
	}
	h := decodeHeapTupleHeader(src[:HeapTupleHeaderSize])
	data := make([]byte, len(src)-HeapTupleHeaderSize)
	copy(data, src[HeapTupleHeaderSize:])
	return &HeapTuple{Header: h, Data: data}, nil
}
